validation: log which peer sent us a header

Sjors opened the pull request:
Fixes #27744

Since #27278 we log received headers. For compact blocks we also log which peer sent it (e5ce8576349d404c466b2f4cab1ca7bf920904b2), but not for regular headers. That required an additional refactor, which this PR provides.

Move the logging from validation to net_processing.

This also reduces the number of log entries (under default configuration) per compact block header from 3 to 2: one for the header and one for the connected tip.

The PR introduces a new helper method `LogBlockHeader`.

When receiving a _compact block_ we call `LogBlockHeader` from the exact same place as where we previously logged. So that log message doesn't change. What does change is that we no longer _also_ log from `AcceptBlockHeader`.

When receiving a regular header(s) message, _we only log the last one_. This is a change in behaviour because it was simpler to implement, but it's probably better anyway. It does mean that if a peer sends of a bunch of headers of which _any_ is invalid, we won't log it (here).

Lastly I expanded the code comment explaining why we log this. It initially only covered selfish mining, but we also care about peers sending us headers but not following up (see e.g. #27626).

Example log:

```
2023-06-05T13:12:21Z Saw new header hash=000000000000000000045910263ef84b575ae3af151865238f1e5c619e69c330 height=792964 peer=0
2023-06-05T13:12:23Z UpdateTip: new best=000000000000000000045910263ef84b575ae3af151865238f1e5c619e69c330 height=792964 version=0x20000000 log2_work=94.223098 tx=848176824 date='2023-06-05T13:11:49Z' progress=1.000000 cache=6.4MiB(54615txo)
2023-06-05T13:14:05Z Saw new cmpctblock header hash=00000000000000000003c6fd4ef2e1246a3f9e1fffab7247344f94cadb9de979 height=792965 peer=0
2023-06-05T13:14:05Z UpdateTip: new best=00000000000000000003c6fd4ef2e1246a3f9e1fffab7247344f94cadb9de979 height=792965 version=0x20000000 log2_work=94.223112 tx=848179461 date='2023-06-05T13:13:58Z' progress=1.000000 cache=7.2MiB(61275txo)
2023-06-05T13:14:41Z Saw new header hash=000000000000000000048e6d69c8399992782d08cb57f5d6cbc81a9f996c3f43 height=792966 peer=8
2023-06-05T13:14:42Z UpdateTip: new best=000000000000000000048e6d69c8399992782d08cb57f5d6cbc81a9f996c3f43 height=792966 version=0x2db3c000 log2_work=94.223126 tx=848182944 date='2023-06-05T13:14:35Z' progress=1.000000 cache=8.0MiB(69837txo)
```

Sjors commented:
cc @jamesob, @instagibbs, @mzumsande

Sjors commented:
There's various ways in which `AcceptBlockHeader` can return `true` or `false` early without setting `state`. This happens before the place where we used to log stuff. That in turn causes f25b8712c368ecfa250a6878af66f3fea3a4bcb0 to potentially log more stuff than it should. Moving to draft while I give it more thought.

Sjors commented:
Added a check to ensure the header was valid and newly seen before we log it. That matches the behaviour from before this PR.

Sjors commented on src/net_processing.cpp:
```diff
@@ -4268,8 +4295,8 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
         }
 
         if (received_new_header) {
```
Since it's possible for `ProcessNewBlockHeaders` to return false without us returning, I wonder if we should only log if it returned `true` (as we do above).

mzumsande replied:
I don't think this is possible though?
The call structure is `ProcessNewBlockHeaders()` -> `AcceptBlockHeader()` -> {`CheckBlockHeader()` / `ContextualCheckBlockHeader()`}, and the latter two functions return either `true` or `state.Invalid(...)`. As a result, `ProcessNewBlockHeaders()` would always return the invalid `state` of the first header that fails, and  `net_processing` would abort.

mzumsande reviewed (commented):
Tested ACK b857d920444241f1598b6be509796d1d8e14dbbb

I like that this makes logging less verbose (by logging one less entry, and by only logging the last header when multiple headers arrive at the same time).

Sjors commented:
I still have to rebase this, will do soon(tm).

Sjors commented:
Rebased!

Sjors commented:
Rebased after #28218 (no longer needs `ActiveChainstate()` to get `IsInitialBlockDownload()`).

Sjors commented:
Added support for `-logips`. This lets you retrieve the IP of a node you disconnected from, without the need for `-debug=net`. This becomes especially useful with #27277 which removes mempool "noise" from `-debug=validation`, so you can e.g. see what the rejection reason was for this header.

Sjors commented:
It turns out there's a third way that we can (and do) receive new headers: unsolicited `BLOCK` messages. I added two new commits that add logging for this, which also slightly refactors the first one.

Since these new changes are a bit more involved, and also touch `CheckBlock()`, I'll hold off on squashing pending opinions… But I do think logging these is very useful.

vasild commented on src/validation.cpp:
```diff
@@ -3974,23 +3974,6 @@ bool ChainstateManager::AcceptBlockHeader(const CBlockHeader& block, BlockValida
-        LogPrintf("%s\n", msg);
-    }
-
     return true;
```
In `master` this logs if `AcceptBlockHeader()` returns `true`. This PR removes the log from here and moves it to the callers.

The call chains are like this:
`ProcessHeadersMessage()` -> `ProcessNewBlockHeaders()` -> `AcceptBlockHeader()` (has the log)
`ProcessMessage()` -> `ProcessNewBlockHeaders()` -> `AcceptBlockHeader()` (has the log)
`submitheader()` -> `ProcessNewBlockHeaders()` -> `AcceptBlockHeader()` (omits the log)
`AcceptBlock()` -> `AcceptBlockHeader()` (omits the log)

are the two log omissions intentional?

Sjors replied:
They're not intentional.

* `submitheader()` is only used for mining, for which this log isn't relevant, so it's fine to omit.
* `AcceptBlock() -> AcceptBlockHeader()`: this is called in the following ways:
  * in `LoadExternalBlockFile()`: not relevant for this log
  * from `ProcessNewBlock()` called from:
    * the kernel demo utility `bitcoin-chainstate.cpp`: not relevant
    * `GenerateBlock()`: (mining / tests)
    * `ProcessBlock()` in net_processing.cpp: 
      * when receiving `CMPCTBLOCK` and in `ProcessCompactBlockTxns()`: part of compact block processing, we've already logged the new header for that earlier
* when receiving a `BLOCK` directly: handled in the third commit

mzumsande replied:
I looked at this independently (tracing the call chains before reading this thread), and I came to the same conclusion as Sjors, except for the case of a direct BLOCK, which has an issue that I mentioned below.

vasild commented on src/net_processing.cpp:
```diff
@@ -1058,6 +1064,8 @@ class PeerManagerImpl final : public PeerManager
 
     void AddAddressKnown(Peer& peer, const CAddress& addr) EXCLUSIVE_LOCKS_REQUIRED(g_msgproc_mutex);
     void PushAddress(Peer& peer, const CAddress& addr) EXCLUSIVE_LOCKS_REQUIRED(g_msgproc_mutex);
+
```
In the commit message of a0edbded5f `validation: log new header from unsolicited block`: `s/only only/only/`

vasild commented on src/net_processing.cpp:
```diff
@@ -944,8 +944,14 @@ class PeerManagerImpl final : public PeerManager
+     * @param[in]   min_pow_checked     The header is known to be part of a sufficiently high-work chain (anti-dos check)
+     * @return                          True if this is a new block
+     */
+    bool ProcessBlock(CNode& node, const std::shared_ptr<const CBlock>& block, bool force_processing, bool min_pow_checked);
```
Would be nice to wrap the long lines in the commit message of 1a5b9d4a09 `validation: log which peer sent us a header` to something like 70-80 columns.

vasild reviewed (commented):
Approach ACK a0edbded5f998c7efb005c8f9a8f9aba9c7bbda9

Sjors commented on src/net_processing.cpp:
```diff
@@ -3345,6 +3359,33 @@ void PeerManagerImpl::ProcessCompactBlockTxns(CNode& pfrom, Peer& peer, const Bl
+    // To prevent log spam, this function should only be called after it was determined that a
+    // header is both new and valid.
+    //
+    // These messages are valuable for detecting potential selfish mining behavior;
+    // if multiple displacing headers are seen near simultaneously across many
+    // nodes in the network, this might be an indication of selfish mining.
```
I'm not sure if it's precisely defined anywhere, but the term is used to describe a situation where a miner intentionally doesn't broadcast their new block and instead waits for a competitor to mine at the same height, hoping that the network picks their side of the fork. When @jamesob first opened an earlier version of this PR, there were rumors of such an attack happening, but this may have been a misunderstanding based on our block connection logging.
https://bitcoin.stackexchange.com/questions/tagged/selfish-mining

Sjors commented on src/net_processing.cpp:
```diff
@@ -1058,6 +1064,8 @@ class PeerManagerImpl final : public PeerManager
     void AddAddressKnown(Peer& peer, const CAddress& addr) EXCLUSIVE_LOCKS_REQUIRED(g_msgproc_mutex);
     void PushAddress(Peer& peer, const CAddress& addr) EXCLUSIVE_LOCKS_REQUIRED(g_msgproc_mutex);
+
+    void LogBlockHeader(uint256 block_hash, std::optional<int> height, const CNode& peer, const bool via_compact_block = false, const bool via_unsolicited_block = false);
```
Fixed

Sjors commented on src/net_processing.cpp:
```diff
@@ -3345,6 +3359,33 @@ void PeerManagerImpl::ProcessCompactBlockTxns(CNode& pfrom, Peer& peer, const Bl
+        hash.ToString(),
+        height ? strprintf(" height=%d", height.value()) : "",
+        peer.GetId(),
+        fLogIPs ? strprintf(" peeraddr=%s", peer.addr.ToStringAddrPort()) : ""
```
Fixed

Sjors commented:
Rebased and addressed @vasild's feedback.

vasild reviewed (approved):
ACK a460dea70a4fb1b6e82534de27b13a4a1f4f06f7

Would be nice if other reviewers assess https://github.com/bitcoin/bitcoin/pull/27826#discussion_r1348849544. I am new to this area of the code.

Sjors commented:
I get between 0 and 2 of these unsolicited blocks per day:

```
2023-11-07T02:19:56.507378Z Saw new header via unsolicited block hash=00000000000000000003310c8a9104bc38473171de3d2628ce608b7418384402 peer=0 peeraddr=x.x.x.x:8333
```

Sjors commented:
Trivial rebase after #28892's fa79a881ce0537e1d74da296a7513730438d2a02.

Sjors commented:
@mzumsande can you take another look?

vasild reviewed (approved):
ACK 708600f3b7df4cf166a2fcdfd3c0dc70b70812f0

As with the previous ACK: would be nice if other reviewers assess https://github.com/bitcoin/bitcoin/pull/27826#discussion_r1348849544.

mzumsande commented on src/net_processing.cpp:
```diff
@@ -3369,10 +3369,11 @@ void PeerManagerImpl::LogBlockHeader(const CBlockIndex& index, const CNode& peer
         via_compact_block ? "cmpctblock " : "",
-        index.GetBlockHash().ToString(),
-        index.nHeight,
+        via_unsolicited_block ? " via unsolicited block" : "",
```
I like the first commit, but the second/third commit are not ideal, because there is a difference between a new block and a new header:
I started up this branch on mainnet while being a few blocks behind (but not that far such that I'd be in IBD so that the log messages would go to `BCLog::VALIDATION`).
Then, the log entries I'd get for each downloaded block ("Saw new header via unsolicited block") are wrong in two ways:
The header is not new (only the block), plus the blocks is not unsolicited (we asked for it with a `getdata` message).

Sjors replied:
That indeed sounds like two bugs, will try to  reproduce.

Sjors replied:
I can reproduce the issue. When the nodes catches up, for a small fraction of blocks it says "Saw new header via unsolicited block". The header refers to a block that was already connected, sometimes more than 25 blocks ago.

Sjors commented:
Rebased to use the new logging functions `LogInfo` and `LogDebug` from #28318 (see `developer-notes.md`).

Sjors commented:
Post cmake rebase, just in case.

Sjors committed:
Log which peer sent us a header

This also supports -logips.

Sjors committed:
Use LogBlockHeader for compact blocks

The only behavior change is that the block height is now added to the log message.

Sjors committed:
Log block header in net_processing

Previously ChainstateManager::AcceptBlockHeader would log when it
saw a new header. This commit moves logging to the call site(s) in
net_processing. The next commits will then log which peer sent it
and whether it was part of a compact block.

This commit changes behavior:
- when multiple headers are received in a single message, only the
  last one is logged
- if any of the headers are invalid, the valid ones are not logged

This happens because net_processing calls ProcessNewBlockHeaders
with multiple headers, which then calls AcceptBlockHeader one
header at a time.

Additionally:
- when the header is received via a compact block, there's no more
  duplicate log (a later commit also unifies logging code paths)

Sjors commented:
Rebased and split into three commits for hopefully easier review.

mzumsande reviewed (commented):
Code Review ACK fcf369df4b37f840e6d2e78643610e47aa14e716

I prefer the newer logging which is less verbose (not every header of a header message with multiple header is logged while catching up with the last hours /days) and includes the peer - this does mean that in some untypical cases (direct BLOCK message with a previously unknown header, submitheader rpc) new headers aren't logged anymore, I don't think that is a major problem.

vasild reviewed (approved):
ACK fcf369df4b37f840e6d2e78643610e47aa14e716

Would be nice to use the new function `LogIP()` (suggested below).

As mentioned already by @mzumsande, new headers via RPC will not be logged anymore. I am not sure whether this deserves a note outside of this PR. A release note? Maybe not.

Sjors committed:
doc: release note for #27826

Sjors commented on src/net_processing.cpp:
```diff
@@ -3371,6 +3380,32 @@ void PeerManagerImpl::ProcessCompactBlockTxns(CNode& pfrom, Peer& peer, const Bl
+        fLogIPs ? strprintf(" peeraddr=%s", peer.addr.ToStringAddrPort()) : ""
+    );
+    if (m_chainman.IsInitialBlockDownload()) {
+        LogDebug(BCLog::VALIDATION, "%s\n", msg);
+    } else {
+        LogInfo("%s\n", msg);
```
That would imply that nobody should ever user the `LogInfo` helper, maybe we should discuss that in a separate issue.

vasild replied:
That's right. It is more broader than this PR, feel free to ignore.

vasild reviewed (approved):
ACK abe43dfadd6325f80975a76aea57a549c3162191

danielabrozzoni commented:
tACK abe43dfadd6325f80975a76aea57a549c3162191

The changes look good to me, although I'm not too familiar with the validation code. I checked (by reading the code, didn't test) that the logs are missing only when receiving a BLOCK message and in some RPC paths, as explained in https://github.com/bitcoin/bitcoin/pull/27826/files#r1348974233
I sketched this diagram to better understand the code and thought I’d share it here in case it helps someone else too :)

![photo_2025-04-28_21-49-02](https://github.com/user-attachments/assets/af4e7c69-1d7e-49c8-85ca-f00ebba9c259)


I did some manual testing on my own node, and was able to verify that cmpctblocks headers aren't logged twice:
```
2025-04-28T18:38:18Z Saw new header hash=00000000000000000001342073dc6156572be7722e3b642e4052d82fcabc580d height=894351 peer=18
2025-04-28T18:38:58Z Saw new cmpctblock header hash=00000000000000000000d140ef281b7ab486209ef8b96e1dbc563ecd4187bc7b height=894352 peer=18
2025-04-28T18:39:49Z Saw new header hash=000000000000000000005128363ae3707f4a812679fb5dceffe98dc77f338e79 height=894353 peer=20
2025-04-28T19:03:58Z Saw new header hash=000000000000000000018078687fa0f17541e5ba492a1f4fa2e07bddcebb36e9 height=894354 peer=15
2025-04-28T19:07:17Z Saw new cmpctblock header hash=00000000000000000001c789ef386c66afce7a0a5040209e32e05379ea1ff801 height=894355 peer=20
2025-04-28T19:10:25Z Saw new cmpctblock header hash=000000000000000000015c57620a0d3bcb14a924f20138e2c9ada14161b07f39 height=894356 peer=20
2025-04-28T19:10:35Z Saw new header hash=00000000000000000001f5ca9b882b3bcaede47925d9e097aa584cb6da14fc37 height=894357 peer=24	
```

Sjors commented:
@mzumsande does that diagram look right to you?

achow101 commented:
ACK abe43dfadd6325f80975a76aea57a549c3162191

mzumsande commented:
> @mzumsande does that diagram look right to you?

Yes - very nice diagram!
